### PR TITLE
remove moab

### PR DESCRIPTION
**Pull request by AriettaZ**
- setting up environment and executing commands based on https://github.com/OSC/moab-ruby/blob/de83cbd83a71184479cad0d1d7da8876b825f744/lib/moab/scheduler.rb#L68-L86

- systemstatus app works the same after changes

screen shot 2018-07-31 at 9 27 55 pm

Solve #58

### CODE REVIEW

File: lib/moab_showq_client.rb [hunk 1]
```diff
@@ -17,24 +17,25 @@ def initialize(cluster)
...
   def setup
-    scheduler = Moab::Scheduler.new(
-      host: @server['host'],
-      lib: @server['lib'],
-      bin: @server['bin'],
-      moabhomedir: @server['homedir']
-    )
+    lib= Pathname.new(@server['lib'].to_s)
+    bin= Pathname.new(@server['bin'].to_s)
+    moabhomedir= Pathname.new(@server['homedir'].to_s)
```
Comment (ericfranz) [inline lib/moab_showq_client.rb]: Since `@server` is an instance var, it might be cleaner code to create separate methods `moab_lib`, `moab_bin`, and `moab_home` that all return pathnames.

Then lines 23-29 could be in a method `showq_summary_xml` and comments saying that it returns the equivalent of `showq -s --xml`.

Then the the setup method is just:

```ruby
def setup
  doc = REXML::Document.new(showq_summary_xml)
  self.active_jobs = doc.root.elements["queue[@option='active']"].attributes["count"].to_i
...
```

**Comment by AriettaZ** [comment]
showq_summary_xml simply returns output right now
```
  def showq_summary_xml
    cmd = moab_bin.join("showq").to_s
    args = ["-s","--host=#{@server['host']}", "--xml"]
    env= {
        "LD_LIBRARY_PATH" => "#{moab_lib}:#{ENV['LD_LIBRARY_PATH']}",
        "MOABHOMEDIR" => "#{moab_home}"
     }.merge(env.to_h)
    o, e, s = Open3.capture3(env, cmd, *args)
    o
  end
```

There's a rescue at the end of setup 
``` 
def setup
    doc = REXML::Document.new(showq_summary_xml)
    ...
rescue
    MoabShowqClientNotAvailable.new
end
```
Do I still need to add a rescue at the end of showq_summary_xml?

```
  def showq_summary_xml
    cmd = moab_bin.join("showq").to_s
    args = ["-s","--host=#{@server['host']}", "--xml"]
    env= {
        "LD_LIBRARY_PATH" => "#{moab_lib}:#{ENV['LD_LIBRARY_PATH']}",
        "MOABHOMEDIR" => "#{moab_home}"
     }.merge(env.to_h)
    o, e, s = Open3.capture3(env, cmd, *args)
    s.success? ? o : raise(CommandFailed, e)
  rescue Errno::ENOENT => e
    raise InvalidCommand, e.message
  end
```

**Comment by AriettaZ** [comment]
The alert is added at the bottom of the index.html.erb. The alert includes cluster name and error message. 
error_message4

File: app.rb [hunk 2]
```diff
@@ -31,6 +32,10 @@ def dashboard_url
   def public_url
      ENV['OOD_PUBLIC_URL'] || "/public"
   end
+  
+  def alert(message)
+      "<div class='row'><div class='alert alert-warning' role='alert'>#{message}</div></div>"
+  end
```
Comment (ericfranz) [inline app.rb]: Should this be an erb instead of a helper? Not sure. Regardless, `message` needs to be HTML escaped.

File: views/index.erb [hunk 3]
```diff
@@ -9,3 +12,5 @@
 	      </a>
       </div>
 <% end %>
+</div>
+<%= alert(@error_msg) if !@error_msg.empty? %>
```
Comment (ericfranz) [inline views/index.erb]: This logic should be moved into the controller. I would create . `@clusters` instance variable in the `get '/clusters' do` section that does something like this:

```ruby
@clusters = CLUSTERS.map { | cluster| MoabShowqClient.new(cluster).setup }
```

Now you have an array of either `MoabShowqClient` or `MoabShowqClientNotAvailable` instances.

1. For each cluster you can render the div.
2. For each cluster you can render the error message.

Also, I noticed that here you are making a human readable form of the `MoabShowqClient` by doing this:

```ruby
@error_msg.concat("#{cluster.metadata.title|| cluster.id.titleize} Cluster: #{(MoabShowqClient.new(cluster).setup).error_message}</br>")
```

So the error message says something like: "Owens Cluster: #{error_message}".

This logic should be in the model too. If `MoabShowqClient#error_message` contains just the actual error message produced, then maybe `MoabShowqClient#friendly_error_message` or `MoabShowqClient#error_message_humanized` would return the string with the prefix, if the error_ message exists. That way you could display an error message for each cluster as appropriate.

File: views/layout.erb [hunk 4]
```diff
@@ -76,6 +76,9 @@
     .jumbotron{
       margin-top: 28px;
     }
+    .alert{
+     margin:20px 0 0 0;   
+    }
```
Comment (ericfranz) [inline views/layout.erb]: Both `.jumbotron` and `.alert` affect all of bootstrap styles. Why is this necessary? Are we breaking a bootstrap convention that requires us to modify the built in styles?

File: views/index.erb (hunk 3 above)
Comment (ericfranz) [reply views/index.erb]: Oh yeah, i realize after seeing your screenshot you are collecting all of the error messages together in one alert. If it makes the code easier, you can just have 3 separate alerts.

**Comment by AriettaZ** [comment]
-  logic moved into controller
    ```
    <%= alert(@error_msg) if !@error_msg.empty? %> 
    ``` 
    from index.erb to app.rb
    ```
    @clusters = CLUSTERS.map { | cluster| MoabShowqClient.new(cluster).setup }
    @error_message = (@clusters.map{ | cluster| cluster.friendly_error_message}).compact 
    ```
- logic moved into model
    ```
    @error_msg.concat("#{cluster.metadata.title|| cluster.id.titleize} Cluster: # 
   {(MoabShowqClient.new(cluster).setup).error_message}</br>")
    ```
    from index.erb to moab_showq_client.rb
    ``` 
      def friendly_error_message
          error_message.nil?? nil : "#{cluster_title} Cluster: #{error_message}"
      end
    ``` 
-  be an erb instead of a helper
    ``` 
      def alert(message)
          "<div class='row'><div class='alert alert-warning' role='alert'>#{message}</div></div>"
      end
    ```
     from app.rb to layout 
    ``` 
      <!-- error message  -->
      <% if @error_message %>
        <% @error_message.each do |msg| %>
            <div class='row'><div class='alert alert-warning' role='alert'><%= msg %></div></div>
        <% end %>
     <% end %><!-- /.error message  -->
    ```
- bootstrap style issue:
  -  `.alert` customized style is removed
  - `.jumbotron` is chanegd to `.jumbotron.error-page`  

- view is changed into separate alert messages
systemstatus_errormessage

File: views/layout.erb [hunk 5]
```diff
@@ -118,9 +118,15 @@
 
 <div class="container" role="main">
   <%= yield %>
-
+  <!-- error message  -->
+  <% if @error_message %>
+    <% @error_message.each do |msg| %>
+        <div class='row'><div class='alert alert-warning' role='alert'><%= msg %></div></div>
+    <% end %>
+ <% end %><!-- /.error message  -->
```
Comment (ericfranz) [inline views/layout.erb]: You can coerce to array to avoid need for if statement:

```
<% Array(@error_message).each do |msg| %>
```

thats because `Array(nil) #=> []` and `Array([1,2,3]) # =>  [1,2,3]`

Comment (ericfranz) [reply views/layout.erb]: And `@some_instance_variable_that_does_not_exist` will resolve to nil:

```
irb(main):004:0> class Foo
irb(main):005:1> def bar
irb(main):006:2> Array(@one)
irb(main):007:2> end
irb(main):008:1> end
=> :bar
irb(main):010:0> Foo.new.bar
=> []
```

Comment (AriettaZ) [reply views/layout.erb]: I changed if statement to `<% Array(@error_messages).each do |msg| %>` and solved small code style issues you mentioned. Tested and all good.
